07-documentation.md

```diff
@@ -0,0 +1,80 @@
+
+Besides an overview of resources the documentation should also discuss cross cutting concerns, such as status codes, error handling and security. The REST document serves as contract, helping consumers of the interface and ensuring that the interface stays consistent and clear.
+
+Writing REST documention takes time, but should be worth it in the process of a project. REST documentation minimizes communication issues between developers and helps new developers become productive in a shorter time span. To really benefit from a REST document it has to stay accurate. Outdated documentation can bring more problems than solutions.
+
+# Tooling
+
+We use tooling to write and maintain the REST documentation in a minimal amount of time. In the sections below two tools are described and compared. Use the tool that best fits your situation.
+
+## Swagger
+
+[Swagger](http://swagger.io/) provides a Spring MVC addon that inspects all REST endpoints and generates an interface description. In combination with [Swagger UI](http://swagger.io/swagger-ui/) a visual test suite can be rendered, showing each controller and request. The requests can be invoked on demand, with parameter and request body support.
+
+Additional documentation can be added by enhancing the controller methods with annotations:
+
+	@ApiResponses(value = {@ApiResponse(code = 405, message = "Invalid input")})
+	public .... createSomething(..)
+
+Swagger is easy to integrate in projects but also has various drawbacks:
+
+- *No cross cutting concerns.* Swagger only shows the requests, there is no way to define cross cutting concerns. One clear example is the usage of error codes. Each request shows a list of all possible error codes, most are clearly irrelevant for that request, but could confuse a frontend developer. The only way to overcome this problem is by heavily annotating your controller with Swagger annotations.
+- *Grouping per controller not resource.* Sometimes a resource is split over various controllers for design reasons. Each controller is shown seperately, while you would rather see them as one resource.
+
+## Spring REST docs
+
+[Spring REST docs](http://docs.spring.io/spring-restdocs/docs/1.0.x/reference/html5/) helps to generate an acurate and readable REST document. Rather than a testsuite we will produce an actual document, using the text processor [Ascii doctor](http://asciidoctor.org/). In this document we can manually describe the cross cutting concerns and resources, allowing us to manage our own markup and chapters.
+
+Frequently changing information, such as HTTP request details, are generated. Spring REST docs generates various snippets from unit tests written in [Spring MVC Test](http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#spring-mvc-test-framework). By using unit tests you ensure that the documentation is actual, otherwise the test will fail. Another nice side effect is that each request will be tested.
```

Comment:
I am confused with this paragraph, you are talking about generating rest documentation, but here you are talking about testing? Is the documentation generated based on the unit tests you write?